### PR TITLE
[Buddy 30860] Added `--insecure` flag to `tbot`

### LINKED ISSUE #20394: Machine ID: '--insecure' flag support
Expected behavior:
tbot ignores the certificate of the auth-proxy.

Current behavior:
tbot errors with an certificate error

Bug details:
- Teleport version: v11.2.2 git:api/v11.2.2-0-g90bde73 go1.19.5

I'm building a local test scenario, I want to run without MFA and use Self Signed certificates.
This has worked fine for Teleport Server and the Agent, but upon starting a bot, it doesn't allow me to connect to the self-signed auth-server.

```
[root@bot ~]# tbot start --destination-dir=./tbot-user --token=<token> --auth-server=teleport:8443
INFO [TBOT]      Created directory "/var/lib/teleport/bot" config/destination_directory.go:134
INFO [TBOT]      Attempting to generate new identity from token tbot/renew.go:455
INFO [AUTH]      Attempting registration with auth server. auth/register.go:245
WARN [AUTH]      Joining cluster without validating the identity of the Auth Server. This may open you up to a Man-In-The-Middle (MITM) attack if an attacker can gain privileged network access. To remedy this, use the CA pin value provided when join token was generated to validate the identity of the Auth Server. auth/register.go:411
INFO [AUTH]      Attempting registration via proxy server. auth/register.go:245
ERROR: WARNING:

  The proxy you are connecting to has presented a certificate signed by a
  unknown authority. This is most likely due to either being presented
  with a self-signed certificate or the certificate was truly signed by an
  authority not known to the client.

  If you know the certificate is self-signed and would like to ignore this
  error use the --insecure flag.

  If you have your own certificate authority that you would like to use to
  validate the certificate chain presented by the proxy, set the
  SSL_CERT_FILE and SSL_CERT_DIR environment variables respectively and try
  again.

  If you think something malicious may be occurring, contact your Teleport
  system administrator to resolve this issue.
```
When using the `--insecure` flag, tbot states it doesn't have one :-)

```
[root@bot ~]# tbot start --insecure --destination-dir=./tbot-user --token=<token> --auth-server=teleport:8443
tbot: error: unknown long flag '--insecure'
```

Either the error shouldn't contain the suggestion, or `tbot` should have an `--insecure` option.

### PR DESCRIPTION
**Pull request by strideynet**
Buddies https://github.com/gravitational/teleport/pull/30860
Closes #20394

---

After filing a bug in #20394, I'd thought, what the hell, let's try to tackle it myself.

What did I do:
- Added the --insecure flag (and yaml entry) to the configuration
- Updated conversion of the CLI Config to a `BotConfig`
- Updated the `registerThroughProxy` function to allow setting the `insecure` flag instead of it just relying on the `lib.IsInsecureDevMode()` function.
- Updated the `testhelpers.DefaultBotConfig` function to allow passing 2 new booleans:
  - `preferAuthServerOverProxyServer` is used for using the Auth server for joining instead of using the Proxy Server. I'm not 100% certain why this is, but according to @strideynet this was required to get the server to properly use the Insecure flag.
  - `insecure` is for passing the Insecure property as a flag instead of implicitly setting it. This should help to set this to false in the future when there would be a full CA Chain used in testing.
- Updated some uses of `TLSConfig` to pass on the Insecure flag from the configuration to the actual TLS Options used.
- Written a Unit test: `TestBot_Insecure`

Extra:
- I've added a `test-go-unit-tbot` `Makefile` target to easily run a subset of unit tests.

I hope I'm doing everything by the book, so please be gentle :)
Feedback is much appreciated!

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -712,6 +712,15 @@ test-go-unit:
 		| tee $(TEST_LOG_DIR)/unit.json \
 		| gotestsum --raw-command -- cat
 
+# Runs tbot unit tests
+.PHONY: test-go-unit-tbot
+test-go-unit-tbot: FLAGS ?= -race -shuffle on
+test-go-unit-tbot: SUBJECT ?= $(shell go list ./... | grep -E '(tool/tbot|lib/tbot)')
```
Comment (zmb3) [inline Makefile:718]:
```suggestion
test-go-unit-tbot: SUBJECT ?= ./tool/tbot/... ./lib/tbot/...
```

File: Makefile [hunk 2]
```diff
@@ -712,6 +712,15 @@ test-go-unit:
...
+test-go-unit-tbot:
+	$(CGOFLAG) go test -cover -json -tags "$(PAM_TAG) $(FIPS_TAG) $(BPF_TAG) $(RDPCLIENT_TAG) $(TOUCHID_TAG) $(PIV_TEST_TAG)" $(PACKAGES) $(SUBJECT) $(FLAGS) $(ADDFLAGS) \
```
Comment (zmb3) [inline Makefile:720]: The `tbot` tests probably don't care about a lot of these tags, right? I can't imagine you need BPF, Touch ID, RDP client, etc.

File: examples/local-cluster/bot/start.sh [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+set -e
+
+SCRIPT_DIR=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+echo "SCRIPT_DIR: ${SCRIPT_DIR}"
```
Comment (zmb3) [inline examples/local-cluster/bot/start.sh:5]: I assume this was only necessary for debugging and can be removed.

File: examples/local-cluster/bot/start.sh [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+echo "Generating bot ID"
+BOT_ID=$(tr -dc a-z0-9 </dev/urandom | head -c 6)
+
+echo "Generating bot token"
+cd ${SCRIPT_DIR}/../auth
+TOKEN=$(tctl -c teleport.yaml bots add static-${BOT_ID} --roles=access --logins=root --format=json | jq .token_id)
+cd -
+
+echo "Resetting config file"
+cp tbot.yaml.orig tbot.yaml
```
Comment (zmb3) [inline examples/local-cluster/bot/start.sh:16]: Can we use `tbot configure` instead of hacking up a file with bash?

File: lib/auth/register.go [hunk 5]
```diff
@@ -611,6 +605,42 @@ func pinRegisterClient(params RegisterParams) (*Client, error) {
...
+	tlsConfig := utils.TLSConfig(params.CipherSuites)
+	tlsConfig.Time = params.Clock.Now
+
+	cert, err := readCA(params.CAPath)
+	if err != nil && !trace.IsNotFound(err) {
+		return nil, trace.Wrap(err)
+	}
+
+	// For backwards compatibility, we fall back to insecure mode if loading the caPath fails.
+	// TODO: At a future date we should remove this
```
Comment (zmb3) [inline lib/auth/register.go:618]: Ideally we have an owner and a rough idea of the future date.
```suggestion
	// DELETE IN 16 (strideynet)
```

Comment (strideynet) [reply lib/auth/register.go:618]: If we're happy to change the behaviour here, it'd definitely be nice to set a date and start warning more aggressively - but I do think we might need to do more investigation as for why it currently behaves the way it does. I really don't understand why the original implementation fails open?

File: Makefile (hunk 2 above)
Comment (strideynet) [reply Makefile:720]: Agreed - we shouldn't need any of these tags bar FIPS.

File: lib/tbot/testhelpers/srv.go [hunk 6]
```diff
@@ -167,16 +175,23 @@ func MakeBot(t *testing.T, client auth.ClientI, name string, roles ...string) *p
...
 	t.Helper()
 
 	authCfg := servicecfg.MakeDefaultConfig()
 	err := config.ApplyFileConfig(fc, authCfg)
 	require.NoError(t, err)
 
+	var authServer = ""
+	if opts.UseAuthServer {
+		authServer = authCfg.AuthServerAddresses()[0].String()
+	} else {
```
Comment (strideynet) [inline lib/tbot/testhelpers/srv.go:189]: It's more concise to write this with the "else" value as the initial value of the variable.
```go
authServer := authCfg.Proxy.WebAddr.String()
if opts.UseAuthServer {
    authServer = authCfg.AuthServerAddresses()[0].String()
}
```

This should resolve the linter complaint as well.

File: lib/tbot/config/config.go [hunk 7]
```diff
@@ -193,6 +196,9 @@ type OnboardingConfig struct {
 	// connection to the Teleport auth server.
 	CAPins []string `yaml:"ca_pins,omitempty"`
 
+	// Insecure instructs `tbot` to trust the Auth Server without verifying the CA.
+	Insecure bool `yaml:"insecure,omitempty"`
```
Comment (strideynet) [inline lib/tbot/config/config.go:200]: This looks unused/added in the wrong place.

File: lib/auth/register.go [hunk 8]
```diff
@@ -371,11 +373,18 @@ func registerThroughAuth(token string, params RegisterParams) (*proto.Certs, err
...
+		client, err = insecureRegisterClient(params)
 	case len(params.CAPins) != 0:
 		client, err = pinRegisterClient(params)
+	case params.CAPath != "":
+		client, err = caPathRegisterClient(params)
 	default:
+		// Falling back to insecure registration.
 		client, err = insecureRegisterClient(params)
 	}
+
```
Comment (strideynet) [inline lib/auth/register.go:387]: This newline can probably be removed.

File: lib/auth/register.go [hunk 9]
```diff
@@ -611,6 +605,42 @@ func pinRegisterClient(params RegisterParams) (*Client, error) {
...
+	}
+
+	// For backwards compatibility, we fall back to insecure mode if loading the caPath fails.
+	// TODO: At a future date we should remove this
+	if trace.IsNotFound(err) {
+		log.Warnf("Falling back to insecurely joining because a missing or empty CA Path was provided.")
+		return insecureRegisterClient(params)
+	}
+
+	certPool := x509.NewCertPool()
```
[the remark this answers is not in the document]
Comment (strideynet) [reply lib/auth/register.go:624]: This currently preserves the existing behaviour.

In this case, we'd be verifying an auth server presenting cert signed by the Teleport host CA. I'm unsure if we've ever made recommendations that users install their host CA into their system roots.

My gut feeling is in this case, the user has explicitly specified a CA certificate file. If I had explicitly specified a CA, I'd expect the system roots to be ignored. It seems less secure to fall back to system roots.

File: lib/tbot/tbot_test.go [hunk 10]
```diff
@@ -394,3 +404,36 @@ func TestBot_ResumeFromStorage(t *testing.T) {
 	thirdBot := New(botConfig, log)
 	require.NoError(t, thirdBot.Run(ctx))
 }
+
+func TestBot_Insecure(t *testing.T) {
```
Comment (strideynet) [inline lib/tbot/tbot_test.go:408]: Nit: We can make it a little clearer what this is testing.

```suggestion
func TestBot_InsecureViaProxy(t *testing.T) {
```

**Review by strideynet — COMMENTED** [review]
I'm generally happy with this and it appears to work from my manual testing. Just a few last "nits".

File: Makefile [hunk 11]
```diff
@@ -712,6 +712,15 @@ test-go-unit:
 		| tee $(TEST_LOG_DIR)/unit.json \
 		| gotestsum --raw-command -- cat
 
+# Runs tbot unit tests
+.PHONY: test-go-unit-tbot
+test-go-unit-tbot: FLAGS ?= -race -shuffle on
+test-go-unit-tbot: SUBJECT ?= ./tool/tbot/... ./lib/tbot/...
+test-go-unit-tbot:
+	$(CGOFLAG) go test -cover -json $(PACKAGES) $(SUBJECT) $(FLAGS) $(ADDFLAGS) \
```
Comment (zmb3) [inline Makefile:720]: What is PACKAGES in this context? Aren't lib/tbot and tool/tbot the packages we're testing?

**Review by zmb3 — APPROVED** [review]

File: Makefile (hunk 11 above)
Comment (strideynet) [reply Makefile]: e7f021ed01

Comment (zmb3) [reply Makefile]: At this point, do we even need `SUBJECT`? Can we just inline the packages right in the `go test` command?

Comment (strideynet) [reply Makefile]: [8f4518e](https://github.com/gravitational/teleport/pull/31093/commits/8f4518e4f7835b2d013ae6e961184c5466de26f5)

**Review by jakule — APPROVED** [review]

**Pull request merged**